source/llassetgen/source/FntWriter.cpp

```diff
@@ -0,0 +1,215 @@
+		}
+		int encoding = int(face->charmap->encoding);
+		for (int i = 0; i < 4; ++i) {
+			int mask = 0xFF;
+			font_info.charset += char(encoding & mask);
+			encoding >>= 8;
+		}
+		std::reverse(font_info.charset.begin(), font_info.charset.end());
+
+		if (face->charmap->encoding == FT_ENCODING_UNICODE) {
+			font_info.use_unicode = true;
+		}
+		else {
+			font_info.use_unicode = false;
+		}
+		// havent found any of the following information
+		// irrelevant for distance fields --> ignore them
+		/*
+		font_info.stretch_h = 1.0f;
+		font_info.use_smoothing = false;
+		font_info.supersampling_level = 1;
+		font_info.padding = { 2.8f, 2.8f, 2.8f, 2.8f };
+		font_info.spacing = { 0, 0 };
+		font_info.outline_thickness = 0;
+		*/
+	}
+
+	void FntWriter::setCharInfo(FT_UInt gindex, Rect<uint32_t> char_area, Vec2<float> offset) {
+		FT_Load_Glyph(face,	gindex, FT_LOAD_DEFAULT);
+		
+		FT_Pos y_bearing = face->glyph->metrics.vertBearingY;
+		max_y_bearing = std::max(y_bearing, max_y_bearing);
+
+		CharInfo char_info;
+		char_info.id = gindex;
+		char_info.x = char_area.position.x;
+		char_info.y = char_area.position.y;
+		char_info.width = char_area.size.x;
+		char_info.height = char_area.size.y;
+		char_info.x_advance = float(face->glyph->linearHoriAdvance) * 16.16f;
+		char_info.x_offset = offset.x;
+		char_info.y_offset = offset.y;
+		char_info.page = 1;
+		char_info.chnl = 15;
+		char_infos.push_back(char_info);
+	}
+
+	void FntWriter::setKerningInfo() {
+		FT_ULong  left_charcode;
+		FT_UInt   left_gindex;
+		left_charcode = FT_Get_First_Char(face, &left_gindex);
+		while (left_gindex != 0) {
+			FT_ULong  right_charcode;
+			FT_UInt   right_gindex;
+			right_charcode = FT_Get_First_Char(face, &right_gindex);
+			while (right_gindex != 0) {
+				FT_Vector kerning_vector;
+				FT_Get_Kerning(face, left_gindex, right_gindex, FT_KERNING_UNSCALED, &kerning_vector);
+				if (kerning_vector.x != 0) {
+					KerningInfo kerning_info;
+					kerning_info.first_id = left_gindex;
+					kerning_info.second_id = right_gindex;
+					// kerning is provided in 26.6 fractions of a pixel
+					kerning_info.kerning = float(kerning_vector.x) / 26.6f;
+					kerning_infos.push_back(kerning_info);
+				}
+				right_charcode = FT_Get_Next_Char(face, right_charcode, &right_gindex);
+			}
+
+			left_charcode = FT_Get_Next_Char(face, left_charcode, &left_gindex);
+		}
+	}
+
+	void FntWriter::readFont() {
+		setFontInfo();
+		setKerningInfo();
+	}
+
+	void FntWriter::setAtlasProperties(Vec2<uint32_t> size, int max_height) {
+		// collect common_info
+		font_common.line_height = max_height;
+		font_common.scale_w = size.x;
+		font_common.scale_h = size.y;
+		font_common.pages = 1;
+		font_common.is_packed = 0;
+	}
+	
+	void FntWriter::SaveFnt(std::string filepath) {
+		font_common.base = max_y_bearing;
+
+		// check for correct file ending
+		if (filepath.substr(filepath.length() - 4) == ".fnt") {
+			filepath = filepath.substr(0, filepath.length() - 4);
+		}
```

Comment:
Checking for correct user input is something the CLI can do, you can assume you're getting a path ending in ".fnt"

Reply:
I think it's nice that the lib checks that, too.
Problem I see here is that there is no "else".

Reply:
Else isn't needed, the code checks if `filepath` ends in ".fnt" and cuts it off so that line 134 doesn't transform it into "file.fnt.fnt".
IMO the library should not care whether the filename ends with ".fnt" or not, it should save the file where it's told to. The applications can make sure that the library is told a file name that makes sense.